search/materialize.go

```diff
@@ -451,7 +407,169 @@ func (m *Materializer) coalescePageRanges(pagedIdx map[int][]RowRange, offset pa
 	return r
 }
 
-type valuesIterator struct {
+// columnIterator iterates through the values of a single column chunk,
+// partitioning reads into batches of pages.
+type columnIterator struct {
+	ctx         context.Context
+	file        *storage.ParquetFile
+	cc          parquet.ColumnChunk
+	pageBatches []pageEntryRead
+
+	iter *pagesBatchIterator
+	err  error
+}
+
+func (c *columnIterator) Next() bool {
+	rv := c.next()
+	if !rv && c.iter != nil {
+		c.iter.release()
+	}
+	return rv
+}
+
+func (c *columnIterator) next() bool {
+	if c.iter == nil || c.err != nil {
+		return false
```

Comment:
In this or, if the iterator is not nill but there was an error could we be leaking resources here by not releasing the iterator during an error?